<a href="https://colab.research.google.com/github/fralfaro/MAT281_2024/blob/main/docs/labs/lab_09.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# MAT281 - Laboratorio N°09

<a id='p1'></a>
## I.- Problema 01


<img src="https://www.svgrepo.com/show/1064/virus.svg" width = "300" align="center"/>

El **cáncer de mama** es una proliferación maligna de células epiteliales en los conductos o lobulillos mamarios. Esta enfermedad clonal surge cuando una célula adquiere, a través de mutaciones, la capacidad de dividirse sin control, formando un tumor que puede invadir tejidos vecinos y propagarse a otras partes del cuerpo.

El conjunto de datos `BC.csv` contiene información sobre pacientes con tumores (benignos o malignos) y diversas características del tumor, calculadas a partir de imágenes digitalizadas de aspirados con aguja fina (FNA) de masas mamarias. Estas características describen los núcleos celulares en la imagen y permiten diferenciar entre tumores benignos y malignos.

A continuación, cargamos el conjunto de datos:

In [1]:
# Importar librerías
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Importar herramientas de Scikit-learn
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.linear_model import LogisticRegression,Ridge
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Configuración de gráficos
%matplotlib inline
sns.set_palette("deep", desat=0.6)
sns.set(rc={'figure.figsize': (11.7, 8.27)})

# Cargar y preparar los datos
df = pd.read_csv("https://raw.githubusercontent.com/fralfaro/MAT281_2024/main/docs/labs/data/BC.csv")
df.set_index('id', inplace=True)

# Transformación de la variable objetivo
df['diagnosis'] = df['diagnosis'].replace({'M': 1, 'B': 0}).astype(int)

# Visualizar las primeras filas del DataFrame
df.head()

<ipython-input-1-99b1ef4544f8>:26: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['diagnosis'] = df['diagnosis'].replace({'M': 1, 'B': 0}).astype(int)


diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
id                                                                          
842302            1        17.99         10.38          122.80     1001.0   
842517            1        20.57         17.77          132.90     1326.0   
84300903          1        19.69         21.25          130.00     1203.0   
84348301          1        11.42         20.38           77.58      386.1   
84358402          1        20.29         14.34          135.10     1297.0   

          smoothness_mean  compactness_mean  concavity_mean  \
id                                                            
842302            0.11840           0.27760          0.3001   
842517            0.08474           0.07864          0.0869   
84300903          0.10960           0.15990          0.1974   
84348301          0.14250           0.28390          0.2414   
84358402          0.10030           0.13280          0.1980   

          concave points_mean  symmetry_mean  ...  radius_worst  \
id                                            ...                 
842302                0.14710         0.2419  ...         25.38   
842517                0.07017         0.1812  ...         24.99   
84300903              0.12790         0.2069  ...         23.57   
84348301              0.10520         0.2597  ...         14.91   
84358402              0.10430         0.1809  ...         22.54   

          texture_worst  perimeter_worst  area_worst  smoothness_worst  \
id                                                                       
842302            17.33           184.60      2019.0            0.1622   
842517            23.41           158.80      1956.0            0.1238   
84300903          25.53           152.50      1709.0            0.1444   
84348301          26.50            98.87       567.7            0.2098   
84358402          16.67           152.20      1575.0            0.1374   

          compactness_worst  concavity_worst  concave points_worst  \
id                                                                   
842302               0.6656           0.7119                0.2654   
842517               0.1866           0.2416                0.1860   
84300903             0.4245           0.4504                0.2430   
84348301             0.8663           0.6869                0.2575   
84358402             0.2050           0.4000                0.1625   

          symmetry_worst  fractal_dimension_worst  
id                                                 
842302            0.4601                  0.11890  
842517            0.2750                  0.08902  
84300903          0.3613                  0.08758  
84348301          0.6638                  0.17300  
84358402          0.2364                  0.07678  

[5 rows x 31 columns]

Con base en la información presentada, responde las siguientes preguntas. Asegúrate de incluir el código necesario para realizar los análisis requeridos y proporciona una breve explicación que describa tus resultados y el proceso que seguiste.




1. Realice un análisis exploratorio del conjunto de datos.

In [8]:
print('Tamaño')
print(df.shape)
print('--------------------------------------------------------------------')
print('Columnas')
print(df.columns)
print('--------------------------------------------------------------------')
print('Tipos de datos')
print(df.info())
print('--------------------------------------------------------------------')
print('Estadistica')
print(df.describe())
print('--------------------------------------------------------------------')
print('Valores unicos')
print(df.nunique())
print('--------------------------------------------------------------------')
print('Valores nulos')
print(df.isna().sum())


Tamaño
(569, 31)
--------------------------------------------------------------------
Columnas
Index(['diagnosis', 'radius_mean', 'texture_mean', 'perimeter_mean',
       'area_mean', 'smoothness_mean', 'compactness_mean', 'concavity_mean',
       'concave points_mean', 'symmetry_mean', 'fractal_dimension_mean',
       'radius_se', 'texture_se', 'perimeter_se', 'area_se', 'smoothness_se',
       'compactness_se', 'concavity_se', 'concave points_se', 'symmetry_se',
       'fractal_dimension_se', 'radius_worst', 'texture_worst',
       'perimeter_worst', 'area_worst', 'smoothness_worst',
       'compactness_worst', 'concavity_worst', 'concave points_worst',
       'symmetry_worst', 'fractal_dimension_worst'],
      dtype='object')
--------------------------------------------------------------------
Tipos de datos
<class 'pandas.core.frame.DataFrame'>
Index: 569 entries, 842302 to 92751
Data columns (total 31 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------     

Hay 569 datos con 31 categorias. Todas las categorias son numeros. No hay valores nulos. En ´diagnosis´ hay solo dos valores, que es lo que se esperaba.

:2. Normalizar las variables numéricas con el método **StandardScaler**.

In [86]:
scaler = StandardScaler()
df_scaled = scaler.fit_transform(df.drop('diagnosis', axis=1))
df_scaled = pd.DataFrame(df_scaled, columns=df.columns[1:], index=df.index)
df_scaled=pd.concat([df['diagnosis'],df_scaled],axis=1).reset_index(drop=True)
df_scaled.head()

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0          1     1.097064     -2.073335        1.269934   0.984375   
1          1     1.829821     -0.353632        1.685955   1.908708   
2          1     1.579888      0.456187        1.566503   1.558884   
3          1    -0.768909      0.253732       -0.592687  -0.764464   
4          1     1.750297     -1.151816        1.776573   1.826229   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0         1.568466          3.283515        2.652874             2.532475   
1        -0.826962         -0.487072       -0.023846             0.548144   
2         0.942210          1.052926        1.363478             2.037231   
3         3.283553          3.402909        1.915897             1.451707   
4         0.280372          0.539340        1.371011             1.428493   

   symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0       2.217515  ...      1.886690      -1.359293         2.303601   
1       0.001392  ...      1.805927      -0.369203         1.535126   
2       0.939685  ...      1.511870      -0.023974         1.347475   
3       2.867383  ...     -0.281464       0.133984        -0.249939   
4      -0.009560  ...      1.298575      -1.466770         1.338539   

   area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0    2.001237          1.307686           2.616665         2.109526   
1    1.890489         -0.375612          -0.430444        -0.146749   
2    1.456285          0.527407           1.082932         0.854974   
3   -0.550021          3.394275           3.893397         1.989588   
4    1.220724          0.220556          -0.313395         0.613179   

   concave points_worst  symmetry_worst  fractal_dimension_worst  
0              2.296076        2.750622                 1.937015  
1              1.087084       -0.243890                 0.281190  
2              1.955000        1.152255                 0.201391  
3              2.175786        6.046041                 4.935010  
4              0.729259       -0.868353                -0.397100  

[5 rows x 31 columns]

3. Realizar un método de reducción de dimensionalidad visto en clases.

In [87]:
pc=PCA()
pc.fit(df_scaled)
explained_variance_ratio=pc.explained_variance_ratio_
cumulative_variance = np.cumsum(explained_variance_ratio)
n_components = np.argmax(cumulative_variance >= 0.95) + 1  #umbral del 95% de varianza
print(f"Número óptimo de componentes: {n_components}")

Número óptimo de componentes: 11


In [90]:
X=df_scaled.drop('diagnosis',axis=1)
y=df_scaled['diagnosis']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
pca=PCA(n_components=11)
X_train_pca = pca.fit_transform(X_train)
X_eval_pca = pca.transform(X_test)

4. Aplique al menos tres modelos de clasificación distintos. Para cada uno de los modelos escogidos, realice una optimización de los hiperparámetros. además, calcule las respectivas métricas. Concluya.

In [99]:
# Logistic Regression
# Optimizacion de hiperparametros
param_grid1={
    'C':[0.001,0.01,0.1,1,10,100],
    'penalty':['l1','l2'],
    'solver':['liblinear','saga'],
    'max_iter':[100,200,300]
}
model1=LogisticRegression()
grid_search1=GridSearchCV(
    estimator=model1,
    param_grid=param_grid1,
    cv=5,
    scoring='accuracy',
    verbose=1,
    n_jobs=-1
)
grid_search1.fit(X_train_pca,y_train)
print("Mejor precisión en validación cruzada:", grid_search1.best_score_)
print("Mejores parámetros:", grid_search1.best_params_)
mejor_modelo1 = grid_search1.best_estimator_
mejor_modelo1.fit(X_train_pca, y_train)
print("Precisión en el conjunto de evaluación: {:.3f}".format(mejor_modelo1.score(X_eval_pca, y_test)))
y_pred1=mejor_modelo1.predict(X_eval_pca)

Fitting 5 folds for each of 72 candidates, totalling 360 fits
Mejor precisión en validación cruzada: 0.9780219780219781
Mejores parámetros: {'C': 0.1, 'max_iter': 100, 'penalty': 'l2', 'solver': 'liblinear'}
Precisión en el conjunto de evaluación: 0.991


Notemos que la precision obtenida luego de escoger el mejor modelo con Grid es mayor que la esperada.

In [102]:
# SVC
# Optimizacion de hiperparametros
param_grid2={
    'C':[0.1,1,10,100],
    'kernel':['linear','rbf','poly'],
    'gamma':['scale','auto',0.1,1,10],
    'degree':[3,4]
}
model2=SVC()
grid_search2=GridSearchCV(
    estimator=model2,
    param_grid=param_grid2,
    cv=5,
    scoring='accuracy',
    verbose=1,
    n_jobs=-1
)
grid_search2.fit(X_train_pca,y_train)
print("Mejor precisión en validación cruzada:", grid_search2.best_score_)
print("Mejores parámetros:", grid_search2.best_params_)
mejor_modelo2 = grid_search2.best_estimator_
mejor_modelo2.fit(X_train_pca, y_train)
print("Precisión en el conjunto de evaluación: {:.3f}".format(mejor_modelo2.score(X_eval_pca, y_test)))
y_pred2=mejor_modelo2.predict(X_eval_pca)

Fitting 5 folds for each of 120 candidates, totalling 600 fits
Mejor precisión en validación cruzada: 0.9780219780219781
Mejores parámetros: {'C': 0.1, 'degree': 3, 'gamma': 'scale', 'kernel': 'linear'}
Precisión en el conjunto de evaluación: 0.991


In [104]:
#Random Forest
# Optimizacion de parametros
param_grid3={
    'n_estimators':[10,50,100,200],
    'max_depth':[10,20,30],
    'min_samples_split':[2,5,10],
    'min_samples_leaf':[1,2,4],
    'max_features':['sqrt','log2']
}
model3=RandomForestClassifier()
grid_search3=GridSearchCV(
    estimator=model3,
    param_grid=param_grid3,
    cv=5,
    scoring='accuracy',
    verbose=1,
    n_jobs=-1
)
grid_search3.fit(X_train_pca,y_train)
print("Mejor precisión en validación cruzada:", grid_search3.best_score_)
print("Mejores parámetros:", grid_search3.best_params_)
mejor_modelo3=grid_search3.best_estimator_
mejor_modelo3.fit(X_train_pca,y_train)
print("Precisión en el conjunto de evaluación: {:.3f}".format(mejor_modelo3.score(X_eval_pca, y_test)))
y_pred3=mejor_modelo3.predict(X_eval_pca)

Fitting 5 folds for each of 216 candidates, totalling 1080 fits
Mejor precisión en validación cruzada: 0.9604395604395604
Mejores parámetros: {'max_depth': 10, 'max_features': 'log2', 'min_samples_leaf': 4, 'min_samples_split': 10, 'n_estimators': 10}
Precisión en el conjunto de evaluación: 0.939


/usr/local/lib/python3.10/dist-packages/numpy/ma/core.py:2820: RuntimeWarning: invalid value encountered in cast
  _data = np.array(data, dtype=dtype, copy=copy,


In [105]:
#Comparacion de metricas
metricas={
    'Modelo':['Logistic Regression','SVC','Random Forest'],
    'Precision':[precision_score(y_test,y_pred1),precision_score(y_test,y_pred2),precision_score(y_test,y_pred3)],
    'Recall':[recall_score(y_test,y_pred1),recall_score(y_test,y_pred2),recall_score(y_test,y_pred3)],
    'F1-Score':[f1_score(y_test,y_pred1),f1_score(y_test,y_pred2),f1_score(y_test,y_pred3)],
    'Accuracy':[accuracy_score(y_test,y_pred1),accuracy_score(y_test,y_pred2),accuracy_score(y_test,y_pred3)]
}
df_metricas=pd.DataFrame(metricas)
df_metricas

Modelo  Precision    Recall  F1-Score  Accuracy
0  Logistic Regression   1.000000  0.976744  0.988235  0.991228
1                  SVC   1.000000  0.976744  0.988235  0.991228
2        Random Forest   0.909091  0.930233  0.919540  0.938596

**Conclusion:** Si bien la precesion en los dos primeros modelos 1, lo que indica un overfitting, la exactitud es mas fiable, por la forma en que se calcula. Segun esta los dos primeros modelos predominan.
Vemos la importancia de estudiar los hiperparametros para conseguir un mejor ajuste en los datos. Aun que sea del orden de 3 decimales, siempre es una mejora.